exe/Main.hs

```diff
@@ -190,22 +190,397 @@ showEvent lock (EventFileDiagnostics (toNormalizedFilePath' -> file) diags) =
+
+    let consultCradle :: Maybe FilePath -> FilePath -> IO (IdeResult HscEnvEq)
+        consultCradle hieYaml cfp = do
+           logInfo logger $ T.pack ("Consulting the cradle for " <> show cfp)
+           cradle <- maybe (loadImplicitCradle $ addTrailingPathSeparator dir) loadCradle hieYaml
+           eopts <- cradleToSessionOpts cradle cfp
+           logDebug logger $ T.pack ("Session loading result: " <> show eopts)
+           case eopts of
+             -- The cradle gave us some options so get to work turning them
+             -- into and HscEnv.
+             Right opts -> do
+               res <- session (hieYaml, toNormalizedFilePath' cfp, opts)
+               return res
+             -- Failure case, either a cradle error or the none cradle
+             Left err -> do
+               dep_info <- getDependencyInfo ([fp | Just fp <- [hieYaml]])
+               let ncfp = toNormalizedFilePath' cfp
+               let res = (map (renderCradleError ncfp) err, Nothing)
+               modifyVar_ fileToFlags $ \var -> do
+                 pure $ Map.insertWith HM.union hieYaml (HM.singleton ncfp (res, dep_info)) var
+               return res
+
+    -- This caches the mapping from hie.yaml + Mod.hs -> [String]
+    let sessionOpts :: (Maybe FilePath, FilePath) -> IO (IdeResult HscEnvEq)
+        sessionOpts (hieYaml, file) = do
+          v <- fromMaybe HM.empty . Map.lookup hieYaml <$> readVar fileToFlags
+          cfp <- liftIO $ canonicalizePath file
+          case HM.lookup (toNormalizedFilePath' cfp) v of
+            Just (opts, old_di) -> do
+              deps_ok <- checkDependencyInfo old_di
+              if not deps_ok
+                then do
+                  -- If the dependencies are out of date then clear both caches and start
+                  -- again.
+                  modifyVar_ fileToFlags (const (return Map.empty))
+                  -- Keep the same name cache
+                  modifyVar_ hscEnvs (return . Map.adjust (\(h, _) -> (h, [])) hieYaml )
+                  consultCradle hieYaml cfp
+                else return opts
+            Nothing -> consultCradle hieYaml cfp
+
+    dummyAs <- async $ return (error "Uninitialised")
+    runningCradle <- newVar dummyAs :: IO (Var (Async (IdeResult HscEnvEq)))
+    -- The main function which gets options for a file. We only want one of these running
+    -- at a time. Therefore the IORef contains the currently running cradle, if we try
+    -- to get some more options then we wait for the currently running action to finish
+    -- before attempting to do so.
+    let getOptions :: FilePath -> IO (IdeResult HscEnvEq)
+        getOptions file = do
+            hieYaml <- cradleLoc file
+            sessionOpts (hieYaml, file)
+    return $ \file -> do
+      liftIO $ join $ modifyVar runningCradle $ \as -> do
+        -- If the cradle is not finished, then wait for it to finish.
+        void $ wait as
+        as <- async $ getOptions file
```

Comment:
IIUC the reason for running the cradle in a background thread is to address #494.

I believe that because of the `alwaysRerun` in the `GhcSession` rule, this async dance is performed once, for every file, for every `shakeRun` that involves an Action that depends on the `GhcSession` rule. So if we want to typecheck a project with 200 files, that's 200 one-shot threads which all they do is probably read off a hashmap. And because of kick, this will happen for every single keystroke...

This worries me, it could be a material performance regressoin for non multi-cradle projects. Have any measurements been done? 
I would prefer if this was doing the simplest thing possible for now, i.e. invoke hie-bios in the current thread, and the fix for #494 left for a future PR.

Reply:
I have used my branch working on GHC regularly, but that doesn't have `kick`. The patch is broken without fixing #494 as you point out. 

We are already spawning thousands of threads of every keystroke because of all the `GetFileExists` rules which go via shake, and that is even worse by default because shake spawns each one onto a new capability (see https://github.com/ndmitchell/shake/pull/751)

So I would prefer not to change this as it already took a bit of effort to get right and doesn't seem to be an issue in practice.

Reply:
Your branch has fundamentally different performance characteristics, so that previous experience doesn't apply. 

Does Shake spawn a new thread for every dependency? I thought it used a threadpool. This change will really spawn one thread for every module, for every keystroke. That includes even simply moving the cursor in some clients like Emacs, which send a hover request per keystroke.

We don't need to change this if the measurements show a non-material performance impact.

Reply:
Yes, shake spawns a new thread for each dependency but bounds the number of threads to the number of capabilities. If you look at an eventlog you will see what happens is that each dependency gets scheduled onto a different capability which is really bad if you have lots of short actions. 

I tried to do some tests just now but GHC master seems broken with ghcide so will have to do it later.

Reply:
Shake could easily avoid spawning fresh threads for each action. I've left a load of comments on https://github.com/ndmitchell/shake/pull/751 - I think the fixes proposed are great, unless the predictions are wrong, and then the whole system becomes single threaded. Happy to get clarification on that ticket.

Reply:
And this code is committing the same sin as Shake, using `async` instead of `asyncOn`. Maybe applying the same trick here would be enough to minimize the cost until we can take some measurements?

Reply:
After testing in the Strats codebase I don't see a huge performance difference, so I am dropping my reservations. What I tested:
- hover (warm)
- diagnostics after a single edit
- hover after edit

Reply:
Thanks Pepe, it is a good observation and I would like set up some more robust telemetry to be able to analyse these issues easily in future.

The difference with shake is that we use an unbounded number of threads so GHC's scheduler can decide how to assign each thread to a capability.